app/models/reports/lsa/fy2021/base.rb

```diff
@@ -0,0 +1,251 @@
+    end
+
+    def missing_data(user)
+      @missing_data = {}
+      @range = ::Filters::DateRange.new(start: Date.current - 3.years, end: Date.current)
+      @missing_data[:missing_housing_type] = missing_housing_types(user)
+      @missing_data[:missing_geocode] = missing_geocodes(user)
+      @missing_data[:missing_geography_type] = geography_types(user)
+      @missing_data[:missing_zip] = missing_zips(user)
+      @missing_data[:missing_operating_start_date] = operating_start_dates(user)
+      @missing_data[:invalid_funders] = invalid_funders(user)
+      @missing_data[:missing_coc_codes] = missing_coc_codes(user)
+      @missing_data[:missing_inventory_coc_codes] = missing_inventory_coc_codes(user)
+      @missing_data[:missing_inventory_start_dates] = missing_inventory_start_dates(user)
+
+      @missing_data[:missing_projects] = @missing_data.values.flatten.uniq.sort_by(&:first)
+      @missing_data[:show_missing_data] = @missing_data[:missing_projects].any?
+      @missing_data
+    end
+
+    private def missing_data_columns
+      {
+        project_name: p_t[:ProjectName].to_sql,
+        org_name: o_t[:OrganizationName].to_sql,
+        project_type: p_t[:computed_project_type].to_sql,
+        funder: f_t[:Funder].to_sql,
+        id: p_t[:id].to_sql,
+        ds_id: p_t[:data_source_id].to_sql,
+      }
+    end
+
+    private def missing_data_rows(scope)
+      scope.pluck(*missing_data_columns.values).
+        map do |row|
+          row = Hash[missing_data_columns.keys.zip(row)]
+          {
+            project: "#{row[:org_name]} - #{row[:project_name]}",
+            project_type: row[:project_type],
+            id: row[:id], data_source_id:
+            row[:ds_id]
+          }
+        end
+    end
+
+    private def missing_housing_types(user)
+      # There are a few required project descriptor fields.  Without these the report won't run cleanly
+      missing_data_rows(
+        GrdaWarehouse::Hud::Project.viewable_by(user).coc_funded.joins(:organization).
+        includes(:funders).
+        where(computed_project_type: [1, 2, 3, 8, 9, 10, 13]).
+        where(HousingType: nil, housing_type_override: nil).
+        where(ProjectID: GrdaWarehouse::Hud::Enrollment.open_during_range(@range).select(:ProjectID)), # this is imperfect, but only look at projects with enrollments open during the past three years
+      )
+    end
+
+    private def missing_geocodes(user)
+      missing_data_rows(
+        GrdaWarehouse::Hud::ProjectCoc.joins(project: :organization).
+        includes(project: :funders).
+        distinct.
+        merge(GrdaWarehouse::Hud::Project.viewable_by(user).coc_funded.hud_residential).
+        where(ProjectID: GrdaWarehouse::Hud::Enrollment.open_during_range(@range).select(:ProjectID)). # this is imperfect, but only look at projects with enrollments open during the past three years
+        where(Geocode: nil, geocode_override: nil),
+      )
+    end
+
+    private def geography_types(user)
```

Comment:
Is there a reason that some of these start with `missing_` and some don't?

Reply:
no reason that I know of.  This was a direct copy from 2020